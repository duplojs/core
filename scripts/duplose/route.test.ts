import { Process } from "./process";
import {
	BuildNoRegisteredDuploseError,
	CutStep,
	Duplose,
	getTypedEntries,
	Hook,
	LastStepMustBeHandlerError,
	OkHttpResponse,
	zod,
	type AnyFunction,
	type Floor,
} from "..";
import { Route } from "./route";
import { Request } from "@scripts/request";
import { PreflightStep } from "@scripts/step/preflight";
import { HandlerStep } from "@scripts/step/handler";
import { Response } from "@scripts/response";
import { CheckpointList } from "@test/utils/checkpointList";
import { DuploTest } from "@test/utils/duploTest";
import { createProcessDefinition, createRouteDefinition } from "@test/utils/manualDuplose";
import { ExtractStep } from "@scripts/step/extract";

describe("Route", () => {
	const checkpointList = new CheckpointList();
	const duplo = new DuploTest({ environment: "TEST" });

	const preflightProcess = new Process(createProcessDefinition());
	preflightProcess.instance = duplo;
	const preflight = new PreflightStep(preflightProcess, { pickup: ["flute"] as any });
	const extractStep = new ExtractStep(
		{
			params: { userId: zod.coerce.number() },
			body: zod.object({
				test: zod.string(),
			}).optional(),
		},
	);
	const cutStep = new CutStep(
		({ dropper }) => {
			checkpointList.addPoint("cut");
			return dropper({ toto: "true" });
		},
		["toto"],
		[new Response(100, "toto", zod.undefined())],
	);

	const route = new Route(createRouteDefinition({
		method: "POST",
		preflightSteps: [preflight],
		steps: [extractStep, cutStep],
	}));

	it("getAllHooks", () => {
		const hooks = route.getAllHooks();

		getTypedEntries(hooks)
			.forEach(([key, value]) => {
				if (!(value instanceof Hook)) {
					return;
				}

				expect(value.subscribers[0])
					.toBe(route.hooks[key]);

				expect((value.subscribers[1] as Hook).subscribers[0])
					.toBe(preflightProcess.hooks[key]);
			});
	});

	it("hasDuplose", () => {
		expect(route.hasDuplose(new Process(createProcessDefinition()))).toBe(false);
		expect(route.hasDuplose(route)).toBe(true);
		expect(route.hasDuplose(preflightProcess)).toBe(true);
	});

	it("build", async() => {
		const spy = vi.spyOn(Duplose.defaultEvaler, "makeFunction");
		checkpointList.reset();

		await expect(() => route.build()).rejects.toThrowError(BuildNoRegisteredDuploseError);

		route.instance = duplo;

		await expect(() => route.build()).rejects.toThrowError(LastStepMustBeHandlerError);

		const handlerStep = new HandlerStep(
			(
				(
					pickup: Floor<{
						toto: string;
						userId: number;
					}>["pickup"],
				) => new OkHttpResponse(pickup("toto"), pickup("userId"))
			) as AnyFunction,
		);
		route.definiton.steps.push(handlerStep);

		await route.build();

		expect(spy).toBeCalled();

		expect(spy.mock.lastCall?.[0].content).toMatchSnapshot();

		route.hooks.onError.addSubscriber((_request, error) => {
			checkpointList.addPoint("onError");
			throw error;
		});

		route.hooks.beforeSend.addSubscriber((_request, response) => {
			checkpointList.addPoint("beforeSend");
			expect(response.code).toBe(200);
			expect(response.information).toBe("true");
			expect(response.body).toBe(2);
		});

		const routeFunction = await route.build();

		await routeFunction(new Request({ params: { userId: "2" } } as any));

		expect(checkpointList.getPointList()).toStrictEqual([
			"start",
			"cut",
			"end",
		]);
	});
});
