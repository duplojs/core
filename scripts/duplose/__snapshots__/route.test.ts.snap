// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Route > build 1`] = `
"
		let result = undefined;

		try {
			let floor = this.makeFloor();

			Execution: {
				
/* hook-beforeRouteExecution-before */
/* end_block */


				result = await this.hooks.beforeRouteExecution(request)
				
				
/* hook-beforeRouteExecution-before-check-result */
/* end_block */


				if(result instanceof this.Response){

break Execution;
}

				
/* hook-beforeRouteExecution-after */
/* end_block */


				
/* preflight-before */
/* end_block */


				
				
/* preflight-process-(0)-before */
/* end_block */


				result = this.preflightSteps[0].processFunction(
					request,
					this.preflightSteps[0].params.options,
					this.preflightSteps[0].params.input(floor.pickup)
				);

				
/* preflight-process-(0)-before-check-result */
/* end_block */


				if(result instanceof this.Response){

break Execution;
}

				
/* preflight-process-(0)-before-drop */
/* end_block */


				floor.drop("flute", result["flute"]);

				
/* preflight-process-(0)-after */
/* end_block */

			

				
/* preflight-after */
/* end_block */


				
			if(request.body === undefined){
				
/* hook-parsingBody-before */
/* end_block */


				result = await this.hooks.parsingBody(request);

				
/* hook-parsingBody-before-check-result */
/* end_block */


				if(result instanceof this.Response){

break Execution;
}

				
/* hook-parsingBody-after */
/* end_block */

			}
			

				
/* steps-before */
/* end_block */


				
/* setp-extract-(0)-before */
/* end_block */


			
/* setp-extract-(0)-(params)-(userId)-before */
/* end_block */

		
			{
				let temp = this.steps[0]
					.extractObject["params"]["userId"]
					.safeParse(request["params"]?.["userId"])
		
				if(!temp.success){
					result = this.steps[0].catchError(
						"params",
						"userId",
						temp.error,
					);
					break Execution;
				}
		
				floor.drop(
					"userId",
					temp.data,
				);
			}
		
			
/* setp-extract-(0)-(params)-(userId)-after */
/* end_block */

			

			
/* setp-extract-(0)-(body)-before */
/* end_block */

		
			{
				let temp = this.steps[0]
					.extractObject["body"]
					.safeParse(request["body"])
		
				if(!temp.success){
					result = this.steps[0].catchError(
						"body",
						undefined,
						temp.error,
					);
					break Execution;
				}
		
				floor.drop(
					"body",
					temp.data,
				);
			}
		
			
/* setp-extract-(0)-(body)-after */
/* end_block */

			

/* setp-extract-(0)-after */
/* end_block */


		
/* step-cut-(1)-before */
/* end_block */


		result = this.steps[1].cutFunction(floor, request);

		
/* step-cut-(1)-before-check-result */
/* end_block */


		if(result instanceof this.Response){

				let temp = this.steps[1].responseZodSchema.safeParse(result);

				if(!temp.success){
					throw new this.ContractResponseError(temp.error, result);
				}
			
break Execution;
}

		
/* step-cut-(1)-before-drop */
/* end_block */


		floor.drop("toto", result["toto"]);

		
/* step-cut-(1)-after */
/* end_block */

		

		
/* step-handler-(2)-before */
/* end_block */


		result = this.steps[2].handlerFunction(floor.pickup, request);

		
/* step-handler-(2)-before-check-result */
/* end_block */


		if(result instanceof this.Response){

break Execution;
}

		
/* step-handler-(2)-after */
/* end_block */

		

				
/* steps-after */
/* end_block */


				
/* defaultResponse-before */
/* end_block */


				result = new this.Response(503, "NO_RESPONSE_SENT", undefined);
			}
		} catch (error) {
			
/* hook-onError-before */
/* end_block */


			result = await this.hooks.onError(request, error) 

			
/* hook-onError-after */
/* end_block */

		}

		
/* check-result-before */
/* end_block */


		if(!(result instanceof this.Response)){
			throw new this.ResultIsNotAResponseError(result)
		}

		
/* check-result-after */
/* end_block */


		return result
		"
`;
